crates/valence/src/instance/chunk.rs

```diff
@@ -46,6 +53,97 @@ struct Section {
     section_updates: Vec<VarLong>,
 }
 
+/// Represents a block with an optional block entity
+#[derive(Clone, Debug)]
+pub struct Block {
+    state: BlockState,
+    /// Nbt of the block entity
+    nbt: Option<Compound>,
+}
+
+impl Block {
+    pub const AIR: Self = Self::new(BlockState::AIR);
+
+    pub const fn new(state: BlockState) -> Self {
+        Self {
+            state,
+            nbt: match state.block_entity_kind() {
+                Some(_) => Some(Compound::new()),
+                None => None,
+            },
+        }
+    }
+
+    /// # Panics
+    /// This function panics if the specified [`BlockState`] does not support a
+    /// block entity yet `nbt` is [`Option::Some`]
+    pub const fn with_nbt(state: BlockState, nbt: Option<Compound>) -> Self {
+        if state.block_entity_kind().is_none() {
+            assert!(nbt.is_none(), "blockstate does not support nbt");
+        }
+        Self { state, nbt }
+    }
+
+    pub const fn state(&self) -> BlockState {
+        self.state
+    }
+}
+
+impl From<BlockState> for Block {
+    fn from(value: BlockState) -> Self {
+        Self::new(value)
+    }
+}
+
+/// Immutable reference to a block in a chunk
+#[derive(Clone, Copy, Debug)]
+pub struct BlockRef<'a> {
+    state: BlockState,
+    nbt: Option<&'a Compound>,
+}
+
+impl<'a> BlockRef<'a> {
+    pub const fn state(&self) -> BlockState {
+        self.state
+    }
+
+    pub const fn nbt(&self) -> Option<&'a Compound> {
+        self.nbt
+    }
+}
+
+/// Mutable reference to a block in a chunk
+#[derive(Debug)]
+pub struct BlockMut<'a> {
+    state: BlockState,
+    /// Entry into the block entity map.
+    entry: Entry<'a, u32, BlockEntity>,
+    modified: &'a mut BTreeSet<u32>,
+}
+
+impl<'a> BlockMut<'a> {
+    pub const fn state(&self) -> BlockState {
+        self.state
+    }
+
+    pub fn nbt_mut(&mut self) -> Option<&mut Compound> {
+        match &mut self.entry {
+            Entry::Vacant(_) => None,
+            Entry::Occupied(oe) => {
+                self.modified.insert(*oe.key());
+                Some(&mut oe.get_mut().nbt)
+            }
+        }
+    }
+
+    pub fn set_nbt(&mut self, nbt: Compound) {
```

Comment:
This can return the previous NBT compound?

Reply:
Actually, I'm not sure this function should exist at all. If you need to mutate the NBT, you can use `nbt_mut` instead and avoid the problems associated with `Block::with_nbt`.